### PR TITLE
tree-wide: enable LTO on Darwin

### PR DESCRIPTION
**Pull request by reckenrode**
## Description of changes

This PR is the second follow-up to https://github.com/NixOS/nixpkgs/pull/307880. It enables LTO in packages flagged by https://github.com/NixOS/nixpkgs/issues/19098 as well as others I could identify as having it disabled on Darwin.

I tested building Python on both aarch64- and x86_64-darwin. I also ran a `nixpkgs-review` on aarch64-darwin with the Python LTO disabled (to keep the number of changed packages manageable). Most packages built. Those that failed did not appear to be LTO-related. I have a number of PRs I will be submitting for some of those broken packages.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by paparodeo** [comment]
<details><summary>enable lto for nuitka</summary>

```
diff --git a/pkgs/development/python-modules/nuitka/darwin-lto.patch b/pkgs/development/python-modules/nuitka/darwin-lto.patch
deleted file mode 100644
index bd3a65905498..000000000000
--- a/pkgs/development/python-modules/nuitka/darwin-lto.patch
+++ /dev/null
@@ -1,15 +0,0 @@
-diff --git a/nuitka/build/SconsCompilerSettings.py b/nuitka/build/SconsCompilerSettings.py
-index 319b72c4e..89d40f2a2 100644
---- a/nuitka/build/SconsCompilerSettings.py
-+++ b/nuitka/build/SconsCompilerSettings.py
-@@ -173,8 +173,8 @@ def _enableLtoSettings(
-         lto_mode = False
-         reason = "known to be not supported (CondaCC)"
-     elif isMacOS() and env.gcc_mode and env.clang_mode:
--        lto_mode = True
--        reason = "known to be supported (macOS clang)"
-+        lto_mode = False
-+        reason = "known to not be supported (macOS nix clang)"
-     elif env.mingw_mode and env.clang_mode:
-         lto_mode = False
-         reason = "known to not be supported (new MinGW64 Clang)"
diff --git a/pkgs/development/python-modules/nuitka/default.nix b/pkgs/development/python-modules/nuitka/default.nix
index fa0842267031..cce4467e73e5 100644
--- a/pkgs/development/python-modules/nuitka/default.nix
+++ b/pkgs/development/python-modules/nuitka/default.nix
@@ -23,9 +23,6 @@ buildPythonPackage rec {
     hash = "sha256-nKdCMgA92v9VsSgfktXDbSh3DyKsGlcTjpn0Y7u4rxU=";
   };
 
-  # default lto off for darwin
-  patches = [ ./darwin-lto.patch ];
-
   build-system = [
     setuptools
     wheel

```
</details>

File: pkgs/development/libraries/wolfssl/default.nix [hunk 1]
```diff
@@ -10,7 +10,7 @@
 # requiring to build a special variant for that software. Example: 'haproxy'
 , variant ? "all"
 , extraConfigureFlags ? []
-, enableLto ? !(stdenv.isDarwin || stdenv.hostPlatform.isStatic || stdenv.cc.isClang)
+, enableLto ? !(stdenv.hostPlatform.isStatic || stdenv.cc.isClang)
```
Comment (paparodeo) [inline pkgs/development/libraries/wolfssl/default.nix:13]: still not enabled

**Comment by reckenrode** [comment]
> enable lto for nuitka

Thanks. I added nuitka to the tree-wide commit.

**Review by paparodeo — APPROVED** [review]
LGTM

File: pkgs/development/libraries/wolfssl/default.nix (hunk 1 above)
Comment (reckenrode) [reply pkgs/development/libraries/wolfssl/default.nix:13]: I don’t know, which is why I left that part of the condition in place. If the clang part is resolved, then Darwin should just work.

**Pull request merged**